### PR TITLE
Fixed 'Add null check' refactoring

### LINKED ISSUE #21501: 'Add null check' refactoring produce incorrect code for Expression-bodied function
**Version Used**: 
VS 2017 15.3 Preview 7.1

**1. Removed trivia**
**Steps to Reproduce**:
```csharp
static void Main(string[] args) /* foo */ => /* foo */
    Console.WriteLine(args.Length);
```

**Expected Behavior**:
```csharp
static void Main(string[] args) /* foo */
{
    if (args == null)
    {
        throw new ArgumentNullException(nameof(args));
    }
    /* foo */
    Console.WriteLine(args.Length);
}
```
**Actual Behavior**:
```csharp
static void Main(string[] args) /* asdasd */
{
    if (args == null)
    {
        throw new ArgumentNullException(nameof(args));
    }

    Console.WriteLine(args.Length);
}
```

**2. No ```return``` statement**
**Steps to Reproduce**:
```csharp
    static bool Foo(string[] args2) =>
        args2.ToList()
            .Select(x => x)
            .Where(x => string.IsNullOrEmpty(x)).Any();
```

**Expected Behavior**:
```csharp
    static bool Foo(string[] args)
    {
        if (args == null)
        {
            throw new ArgumentNullException(nameof(args));
        }

        return args.ToList()
            .Select(x => x)
            .Where(x => string.IsNullOrEmpty(x)).Any();
    }
```
**Actual Behavior**:
```csharp
    static bool Foo(string[] args)
    {
        if (args == null)
        {
            throw new ArgumentNullException(nameof(args));
        }

        args.ToList()
            .Select(x => x)
            .Where(x => string.IsNullOrEmpty(x)).Any();
    }
```

**Steps to Reproduce**:
```csharp
    public int Foo(int[] array) =>
#if DEBUG
            array.Where(x => x > 3)
            .OrderBy(x => x)
            .Count();
#else
            array.Where(x => x > 3)
                .OrderBy(x => x)
                .Count();
#endif
```
**Expected Behavior**:
We shouldn't offer refactoring, same as for 'User block body for methods' refactoring.

**Actual Behavior**:
```csharp
    public int Foo2(int[] array)
    {
        if (array == null)
        {
            throw new ArgumentNullException(nameof(array));
        }
#if DEBUG
        array.Where(x => x > 3)
            .OrderBy(x => x)
            .Count();
    }
#else
            array.Where(x => x > 3)
                .OrderBy(x => x)
                .Count();
#endif
```

### PR DESCRIPTION
**Pull request by zaytsev-victor**
Fixes #21501

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs [hunk 1]
```diff
@@ -888,5 +888,121 @@ public C(string s)
...
+        public async Task TestInArrowExpression1()
+        {
+            await TestInRegularAndScript1Async(
+@"
+using System;
+using System.Linq;
+
+class C
+{
+    public int Foo(int[] array[||]) =>
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: Can you add a test where it's a void method?

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs [hunk 2]
```diff
@@ -30,5 +32,15 @@ protected override void InsertStatement(SyntaxEditor editor, BaseMethodDeclarati
 
         protected override bool IsImplicitConversion(Compilation compilation, ITypeSymbol source, ITypeSymbol destination)
             => InitializeParameterHelpers.IsImplicitConversion(compilation, source, destination);
+
+        protected override bool CanOffer(SyntaxNode blockStatement)
+        {
+            if (blockStatement is ArrowExpressionClauseSyntax arrowExpressionClauseSyntax)
+            {
+                return arrowExpressionClauseSyntax.TryConvertToStatement(SyntaxFactory.Token(SyntaxKind.SemicolonToken), false, out var _);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: named parameters for literals.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs [hunk 3]
```diff
@@ -30,5 +30,18 @@ protected override void InsertStatement(SyntaxEditor editor, BaseMethodDeclarati
 
         protected override bool IsImplicitConversion(Compilation compilation, ITypeSymbol source, ITypeSymbol destination)
             => InitializeParameterHelpers.IsImplicitConversion(compilation, source, destination);
+
+        protected override bool CanOffer(SyntaxNode blockStatement)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpAddParameterCheckCodeRefactoringProvider.cs]: nit: we should rename 'blockStatement' to 'body'.

**Comment by CyrusNajmabadi** [comment]
Awesome!

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs (hunk 1 above)
Comment (zaytsev-victor) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: Done.

**Comment by zaytsev-victor** [comment]
Any update on this?

**Comment by CyrusNajmabadi** [comment]
@dpoeschl @sharwell Can someone merge this?  Thanks!

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs [hunk 4]
```diff
@@ -888,5 +888,157 @@ public C(string s)
...
+        if (array == null)
+        {
+            throw new ArgumentNullException(nameof(array));
+        }
+
+        array.Where(x => x > 3)
+            .OrderBy(x => x)
+            .Count();
+    }
+}", ignoreTrivia: false);
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: :memo: `ignoreTrivia` is now the default :smile:

Comment (sharwell) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: :memo: To be clear here, `ignoreTrivia` now defaults to *false*, meaning trivia will be compared during tests.

Comment (sharwell) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs]: @zaytsev-victor Actually it looks like the option was later removed altogether in #21745, so the build will fail until these uses are updated.

**Pull request merged**